api/ondemand_vhds_rds_protocol.rst

```diff
@@ -0,0 +1,187 @@
+.. _on_demand_vhds_rds_protocol:
+
+On-Demand RDS/VHDS
+==================
+Currently, in RDS, all routes for the cluster are sent to every Envoy instance
+in the mesh. This causes scaling issues as the size of the cluster grows. The
+majority of this complexity can be found in the virtual host configurations, of
+which most are not needed by any individual proxy. With a goal of being able
+to scale to one million vhosts in the future, the capability to filter a
+proxy to only contain the required virtual hosts is a necessity.
+
+In order to fix this issue, we are implementing on-demand RDS. On-demand RDS,
+uses the delta xDS protocol, which is a separate protocol xDS. Instead of
+always sending the full route config, Envoy will be able to subscribe/
+unsubscribe from a list of resources stored internally in the xDS management
+server. The xDS management server will monitor this list and use it to filter
+the configuration sent to an individual Envoy instance to only contain the
+subscribed resources.
+
+Joining/Reconnecting
+====================
+When an Envoy instance forms a delta xDS connection to the xDS management
+server for the first time, it will receive a base configuration filtered down
+to a subset of routes that are likely useful to the Envoy instance. For
+example, these might be filtered down by routes that exist in the current
+namespace. The xDS management server will send this same collection of
+resources on reconnect.
+
+.. figure:: diagrams/delta_rds_connection.svg
+   :alt: Delta RDS connection
+
+Requesting Additional resources
+===============================
+In order to request additional resources, Envoy will send a
+:ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`
+to a xDS management server. The xDS management server will then respond with a
+:ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>`
+which is followed up by an Ack/Nack
+:ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>`
+from Envoy.
+
+.. figure:: diagrams/delta_rds_request_additional_resources.svg
+   :alt: Delta RDS request additional resources
+
+Virtual Host Discovery Service
```

Comment:
Yeah, I'm a bit confused why we have separate RDS and VHDS treatment here. I think that VHDS is the main contribution, and RDS hasn't changed much (excepting the SRDS work).

Reply:
RDS has changed such that the virtual host list is no longer populated and instead requested on-demand instead.

Reply:
> RDS has changed such that the virtual host list is no longer populated and instead requested on-demand instead.

It may or may not be populated, current VHDS implementation handles both cases.

Reply:
+1